raphtory-graphql/src/data.rs

```diff
@@ -1,68 +1,85 @@
 };
-use std::{collections::HashMap, fs, path::Path, sync::Arc};
+use std::{
+    collections::HashMap,
+    fs,
+    path::{Path, PathBuf},
+    sync::Arc,
+};
 use walkdir::WalkDir;
 
-#[derive(Default)]
 pub struct Data {
-    pub(crate) graphs: Arc<RwLock<HashMap<String, IndexedGraph<MaterializedGraph>>>>,
-    pub(crate) vector_stores: Arc<RwLock<HashMap<String, DynamicVectorisedGraph>>>,
+    pub(crate) work_dir: String,
+    pub(crate) graphs: Arc<Cache<String, IndexedGraph<MaterializedGraph>>>,
+    pub(crate) global_plugins: GlobalPlugins,
 }
 
 impl Data {
-    pub fn from_map<G: Into<MaterializedGraph>>(graphs: HashMap<String, G>) -> Self {
-        let graphs = Arc::new(RwLock::new(Self::convert_graphs(graphs)));
-        let vector_stores = Arc::new(RwLock::new(HashMap::new()));
-        Self {
-            graphs,
-            vector_stores,
-        }
-    }
+    pub fn new(
+        work_dir: &Path,
+        graphs: Option<HashMap<String, MaterializedGraph>>,
+        graph_paths: Option<Vec<PathBuf>>,
+        configs: &AppConfig,
+    ) -> Self {
+        let cache_configs = &configs.cache;
+
+        let graphs_cache_builder = Cache::builder()
+            .max_capacity(cache_configs.capacity)
+            .time_to_idle(std::time::Duration::from_secs(cache_configs.tti_seconds))
+            .build();
+
+        let graphs_cache: Arc<Cache<String, IndexedGraph<MaterializedGraph>>> =
+            Arc::new(graphs_cache_builder);
+
+        save_graphs_to_work_dir(work_dir, &graphs.unwrap_or_default())
+            .expect("Failed to save graphs to work dir");
+
+        load_graphs_from_paths(work_dir, graph_paths.unwrap_or_default(), true)
+            .expect("Failed to save graph paths to work dir");
 
-    pub fn from_directory(directory_path: &str) -> Self {
-        let graphs = Arc::new(RwLock::new(Self::load_from_file(directory_path)));
-        let vector_stores = Arc::new(RwLock::new(HashMap::new()));
         Self {
-            graphs,
-            vector_stores,
+            work_dir: work_dir.to_string_lossy().into_owned(),
+            graphs: graphs_cache,
+            global_plugins: GlobalPlugins::default(),
         }
     }
 
-    pub fn from_map_and_directory<G: Into<MaterializedGraph>>(
-        graphs: HashMap<String, G>,
-        directory_path: &str,
-    ) -> Self {
-        let mut graphs = Self::convert_graphs(graphs);
-        graphs.extend(Self::load_from_file(directory_path));
-        let graphs = Arc::new(RwLock::new(graphs));
-        let vector_stores = Arc::new(RwLock::new(HashMap::new()));
-        Self {
-            graphs,
-            vector_stores,
+    pub fn get_graph(&self, name: &str) -> Result<Option<IndexedGraph<MaterializedGraph>>> {
+        let name = name.to_string();
+        let path = Path::new(&self.work_dir).join(name.as_str());
+        if !path.exists() {
+            Ok(None)
+        } else {
+            match self.graphs.get(&name) {
```

Comment:
I think you want to call `get_with(&name, || get_graph_from_path..)` here this will make sure you only load the graph once when called by multiple threads